### PR TITLE
(PUP-5109) Add :serverip6 fact

### PR DESCRIPTION
**Pull request by kris-bosland**
If there is no IPv4 address, use the IPv6 address if available.

### CODE REVIEW

**Comment by joshcooper** [comment]
Looks good, two questions:

* Should this target master for IPv6 work?
* Should the `serverip` server fact contain either IPv4 or v6? Or should there be a separate `serveripv6` server facts?

If it's the former, then constructions like this won't work:

```puppet
file { '/path/to/file':
  ensure => file,
  source => "puppet://$server_fact['serverip']:8140/custom-mount/path/to/file"
}
```

because literal IPv6 addresses in URLs need to be wrapped in brackets:

```
      http://[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:80/index.html
```

On the other hand, if we have 2 server facts (`serverip` and `serveripv6`) then manifests would need to have conditional logic:

```puppet
$host = $server_facts['serverip']
if $host {
  $url = "https://${host}:8140/custom-mount/file"
} else {
  $hostv6 = $server_facts['serveripv6']
  # note brackets
  $url = "https://[${hostv6}]:8140/custom-mount/file"
}

file { '/path/to':
  ensure => file,
  source => $url,
}
```

Or we could have `server_facts['serverip']` return an IPv6 address wrapped in brackets, so that   `source => "puppet://$server_fact['serverip']:8140/custom-mount/path/to/file"` just works?

**Comment by b4ldr** [comment]
> Or we could have server_facts['serverip']
please don't do this as nearly everything else will expect the address without the surrounding []

File: spec/unit/indirector/catalog/compiler_spec.rb [hunk 1]
```diff
@@ -414,6 +426,25 @@ def a_request_that_contains(facts)
...
+        'fqdn'       => "my.server.com",
+        'ipaddress'  => nil,
+        'ipaddress6' => "my.ipv6.address"
+        })
+      @request = Puppet::Indirector::Request.new(:catalog, :find, node_name, nil)
+      allow(compiler).to receive(:compile)
+      allow(Puppet::Node.indirection).to receive(:find).with(node_name, anything).and_return(node)
+    end
+
+    it "should use the IPv6 address for serverip if there is no IPv4 address" do
```
Comment (joshcooper) [inline spec/unit/indirector/catalog/compiler_spec.rb]: minor nit, we decided not to populate `serverip` with the IPv6, so this description needs updating

**Comment by jtappa** [comment]
@kris-bosland to follow up with another PR removing the warning/only warning if there are no addresses resolved.

**Pull request merged**